### DISCUSSION TITLE
Retrieving sidebar/views state

### DISCUSSION THREAD

**Discussion by charisk** [microsoft/vscode-discussions#840] [Extension Development QnA] [answered]
In a custom VS Code extension, is it possible to query the state of sidebars/views (primary or secondary) to find out:
- On which sidebar a specific view is on (a custom view contributed by the extension)?
- Whether a sidebar is opened or closed?

**Comment by sketchbuch** [comment]
Webviews have a visible property and there is an event you can subscribe to to receive notification if your webview's visibility changes (onDidChangeVisibility).

**Reply by charisk** [reply]
I don't think this answers my questions unfortunately. I don't need to know if my view is visible, I need to know if a sidebar is visible.

**Comment by theReynald** [comment, marked as answer]
From my research, there isn't a direct and comprehensive API in VS Code to query the complete state of sidebars/views but VS Code API does expose some information and functionality regarding the view and its visibility, but it's limited. 

Here's what you can try:

1. On which sidebar a specific view is on?
    - Currently, there isn't a direct API to query which sidebar a view is in. You might be able to infer some information based on the view's context (e.g., whether it's in the explorer, SCM, etc.) but there's no direct query to ask, "In which sidebar is this view?".
    - If this is essential for your extension, one way could be to use `WorkspaceConfiguration` to monitor changes in the user's settings, specifically the `"workbench.layout.explorerPosition"` setting, but this only refers to the explorer's position and doesn't guarantee the view's exact position.

2. Whether a sidebar is opened or closed?
    - For views that you contribute, you can use the `onDidChangeVisibility` event from `vscode.window.createTreeView`. This will notify you when your view is hidden or shown, which can give you a hint about the sidebar's state (at least in terms of visibility of your view).
    
    ```typescript
    const treeView = vscode.window.createTreeView('yourTreeViewId', { treeDataProvider: yourDataProvider });
    treeView.onDidChangeVisibility(e => {
        if (e.visible) {
            // Your view is visible
        } else {
            // Your view is hidden
        }
    });
    ```

**Reply by charisk** [reply]
Thanks @theReynald! I think this answers my question - it doesn't look like it's possible to achieve what I'm after exactly, but can infer a few things from  `WorkspaceConfiguration`.

**Comment by theReynald** [comment]
Have you tried to determine if a sidebar is visible or not using`vscode.workspace.getConfiguration` method to access the configuration settings of VS Code? Specifically, you can check the settings related to the visibility of specific sidebars.

In VS Code, different sidebars (e.g., Explorer, Source Control, Extensions) can be shown or hidden based on user preferences. You can access these preferences in your extension code.

Example of how to check if the Explorer sidebar is open or closed:

   ```javascript
   const vscode = require('vscode');

   const config = vscode.workspace.getConfiguration('explorer');
   const isExplorerSidebarVisible = config.get('autoReveal') === true;

   if (isExplorerSidebarVisible) {
       // The Explorer sidebar is open.
   } else {
       // The Explorer sidebar is closed.
   }
   ```

   You can replace `'explorer'` with the configuration key of the specific sidebar you want to check.

**Comment by r3m0t** [comment]
You could try `await vscode.commands.executeCommand("getContextKeyInfo")` and looking for a related context key.
